docs: Add reference for reset and boot sequence, document factory reset for esp32 & rp2.

projectgus opened the pull request:
### Summary

This started out as a task to document `boot.py` and `main.py` on the rp2 port (Closes #15230). However, I realised the boot sequence and reset types were documented per-port and sometimes inconsistent so I decided to consolidate these into a common reference page that each port can link to.

In follow-up commits I added port-specific factory reset docs for esp32 and rp2 ports. These ports don't have any MicroPython-specific recovery features like Safe Mode boots, the factory reset process is to erase the entire flash. However, now it's explicitly documented as such.

projectgus committed:
docs: Add a "Reset and Boot Sequence" reference page.

Previously individual ports documented these aspects to varying degrees,
but most of the information is common to all ports.

In particular, this adds a canonical explanation of `boot.py` and
`main.py`.

This work was funded through GitHub Sponsors.

Signed-off-by: Angus Gratton <angus@redyak.com.au>

projectgus committed:
docs/rp2: Add a small factory reset page.

This work was funded through GitHub Sponsors.

Signed-off-by: Angus Gratton <angus@redyak.com.au>

projectgus committed:
docs/esp32: Add a factory reset page.

This work was funded through GitHub Sponsors.

Signed-off-by: Angus Gratton <angus@redyak.com.au>

projectgus commented:
@TradigK Interested if you have any comments on this PR. Does it look like it resolves your comment from here: https://github.com/micropython/micropython/issues/15230#issuecomment-2328175797 ?

TRadigk commented:
@projectgus  I've checked all changes and fully agree with them. The boot sequence is clearly laid out and to be found on a top level 👍 (which helps to understand that this is not port port specific, but micropython specific). You added a way to recover from soft-bricking for every port allowing it to be found with the search and providing guidance to recover, which I also like very much.
I think this is a great addition to the docs! Thank you! 😎

projectgus committed:
docs/library: Note link between machine.soft_reset() and sys.exit().

This is currently an implementation detail of MicroPython rather than by
design.

This work was funded through GitHub Sponsors.

Signed-off-by: Angus Gratton <angus@redyak.com.au>

projectgus commented:
After discussion with @dpgeorge have added a commit which deprecates `machine.soft_reset()` and recommends `sys.exit()` instead.

dpgeorge commented on docs/reference/reset_boot.rst:
```diff
@@ -0,0 +1,216 @@
+startup process.
+
+``main.py`` is for any Python code that you want to run each time your device
+starts.
```
It might be worth adding somewhere in this file that `main.py` execution is skipped when you soft-reset in raw-repl mode (see docs/reference/repl.rst for more details about raw-repl).

(I'm surprised we don't already document that fact somewhere... at least I couldn't find it.)

dpgeorge commented on docs/esp32/tutorial/reset.rst:
```diff
@@ -0,0 +1,25 @@
+       esptool.py -p PORTNAME erase_flash
+
+3. Use esptool to flash the MicroPython firmware to your board again. If needed,
+   this file and flashing instructions can be found on the `MicroPython
```
file -> firmware

projectgus replied:
Have changed, but would like to note that this term was adopted from the linked esp32 tutorial which uses it extensively, for example "And then deploy the new firmware using::". (I did change "deploy" to "flash", though.)

(Can change the tutorial as well, but suggest I do it in the PR mentioned above.)

dpgeorge commented on docs/library/pyb.rst:
```diff
@@ -205,8 +205,9 @@ Miscellaneous functions
-   Set the filename of the main script to run after boot.py is finished.  If
-   this function is not called then the default file main.py will be executed.
+   Set the filename of the main script to run after :ref:`boot_py` is finished.
+   If this function is not called then the default file main.py will be
```
Maybe also make `main.py` here a ref-link?

andrewleech commented on docs/reference/reset_boot.rst:
```diff
@@ -0,0 +1,216 @@
+
+A file named ``boot.py`` can be copied to the board's internal :ref:`filesystem
+<filesystem>` using :doc:`mpremote <mpremote>` or :doc:`pyboard.py
+<pyboard.py>`.
```
Can they be both frozen on all ports now? I know it only worked on a few ports previously

projectgus replied:
> Also might be worth mentioning that boot.py (and main.py below) can be frozen into the firmware.

The note about freezing is in the list of `main.py` tips (but it mentions you can freeze both files.)

> Can they be both frozen on all ports now? I know it only worked on a few ports previously

Oh..? I've used the phrase "on most ports" and "and on some ports" in other places, should I add it here?

(This is frustratingly non-specific for users, but it's still useful to document where something may be port-dependent than not to document it at all.)

dpgeorge replied:
It looks like all ports now use `pyexec_file_if_exists()` to run `boot.py` and `main.py`, and if freezing is enabled on that port then that function will first search frozen code.

So, all ports now support freezing of these two.

projectgus replied:
Thanks for checking this, @dpgeorge

andrewleech commented on docs/esp32/tutorial/reset.rst:
```diff
@@ -0,0 +1,25 @@
+2. Find the serial port name of your board, and then use esptool to erase the
+   entire flash contents::
+
+       esptool.py -p PORTNAME erase_flash
```
For most users these days the auto-detection of port works well - it's often only advanced users that have more than one com port on their computer at any one time. As such I'd remove the need to "find serial port name" as new users wont know how to do this. 
If desired this could be split to show the basic `esptool.py erase_flash` first and then a short follow up paragraph on setting the port name as well if needed?

projectgus replied:
*Tries to imagine a world where there aren't nine serial devices plugged into a computer at all times*
![image](https://github.com/user-attachments/assets/66a4372e-a788-406d-96fd-bbcee2a6593e)

:grin: In all seriousness, I think I agree. However, I suggest I submit a separate PR for this that also changes it in the esp32 tutorial. Also there are the download page instructions, although I'm not sure where these are generated from - maybe damien has to make that change somewhere internally.

dpgeorge replied:
> Also there are the download page instructions, although I'm not sure where these are generated from

They should be in the various `deploy.md` files (part of the `board.json` and `board.md` set of files).

andrewleech commented on docs/reference/reset_boot.rst:
```diff
@@ -0,0 +1,216 @@
+Hard reset
+----------
+
+Booting from hard reset happens when the device is powered on, if Python code
+executes :func:`machine.reset()`, or if a physical Reset button is pressed on
+the board (where applicable).
```
```suggestion

Booting from hard reset is what happens when a board is first powered up, a cold boot, this is a complete reset of the MCU. 

The same hard reset boot sequence can be triggered by a number of other sources such as:
* Python code executing :func:`machine.reset()`
* User presses a physical Reset button on the board (where applicable)
* MCU watchdog reset
* Waking from deep sleep (on most ports)
* MCU brown out detector (on some ports)
```

andrewleech commented on docs/reference/reset_boot.rst:
```diff
@@ -0,0 +1,216 @@
+
+   The purpose of this late initialisation is so that it's possible to
+   pre-configure particular hardware in ``boot.py``, and then have it start with
+   the correct configuration.
```
It might be also worth mentioning it's common to configure network / wifi in boot.py though as that means it will be available on all boots for use with repl / mpremote when there's no main.py used.

andrewleech reviewed (commented):
These changes look great thanks for writing them!

projectgus commented:
@andrewleech @dpgeorge Thanks for the great reviews! I've made some edits based on your suggestions, please take a look. (Apologies, I rebased so the "compare" feature in GitHub may show a lot of unrelated edits.)

dpgeorge commented:
I have reviewed the changes here and am happy with this PR now.

Except the commit that deprecates `machine.soft_reset()` in favour of `sys.exit()`, that needs reconsideration in the light of https://github.com/micropython/micropython/pull/15486#issuecomment-2445553809

projectgus commented:
After discussion with @dpgeorge, `machine.soft_reset()` is no longer deprecated as the `sys.exit()` behaviour may change in the future. Updated the relevant commit in this PR to document the current behaviour of `sys.exit()`.

projectgus commented:
Also rewrote the suggested `try/except/finally` structure for main.py as the original version locked the user out of the REPL (reset in finally block). This version allows KeyboardInterrupt to drop to a REPL, a safer recommendation...

dpgeorge commented:
> Also rewrote the suggested `try/except/finally` structure for main.py as the original version locked the user out of the REPL (reset in finally block). This version allows KeyboardInterrupt to drop to a REPL, a safer recommendation...

It's tricky to recommend something because in production you'd usually want the app to be as robust and secure as possible and hence not allow ctrl-C or anything to stop it, but that contradicts with having a convenient development cycle, eg using `mpremote`.

Anyway, what you have now is good!
